lib/class-wp-rest-menu-items-batch-processor.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/**
+ * Class to batch-process multiple menu items in a single request
+ *
+ * @see WP_REST_Posts_Controller
+ */
+class WP_REST_Menu_Items_Batch_Processor {
+
+	private $controller;
+
+	public function __construct( WP_REST_Menu_Items_Controller $controller ) {
+		global $wpdb;
+
+		$this->controller = $controller;
+		$this->wpdb = $wpdb;
+	}
+
+	public function process( $navigation_id, $tree ) {
+		$this->controller->ignore_position_collision = true;
+
+		$current_menu_items = wp_get_nav_menu_items( $navigation_id, array( 'post_status' => 'publish,draft' ) );
+
+		$result = $this->wpdb->query('START TRANSACTION');
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+		$stack = [
+			[ 0, $tree ],
+		];
+		$updated_ids = [];
+		while ( ! empty( $stack ) ) {
+			list( $parent_id, $raw_menu_items ) = array_pop( $stack );
+			foreach ( $raw_menu_items as $raw_menu_item ) {
+				$raw_menu_item['parent'] = $parent_id;
+				$children = ! empty( $raw_menu_item['children'] ) ? $raw_menu_item['children'] : [];
+				unset( $raw_menu_item['children'] );
+
+				if ( $raw_menu_item['id'] ) {
+					$updated_ids[] = $raw_menu_item['id'];
+					$result = $this->request_update( $raw_menu_item );
+				} else {
+					$result = $this->request_insert( $raw_menu_item );
+				}
+				if ( is_wp_error( $result ) ) {
+					$this->wpdb->query('ROLLBACK');
+					return $result;
+				}
+				$database_id = $result['id'];
+				if ( $children ) {
+					array_push( $stack, [ $database_id, $children ] );
+				}
+			}
+		}
+
+		// Delete any orphaned items
+		foreach ( $current_menu_items as $item ) {
+			if ( ! in_array( $item->ID, $updated_ids ) ) {
+				$this->request_delete( $item->ID );
+			}
+		}
+		$result = $this->wpdb->query('COMMIT');
```

Comment:
What about if the request times out, what happens? How about all caches that were updated? How are these rolled back?

Reply:
@spacedmonkey still exploring - with the current "brute force" way it takes 40s to save 500 menu items which isn't terrible - I think I can get it down to 2-5 with some tweaks.

> How about all caches that were updated? How are these rolled back?

Good question - I'm not sure yet, I can imagine a scenario where it's so non-trivial we'll have to abandon this approach for the time being. I have some ideas like disabling any hooks, performing SQL queries, and only notifying them if everything worked properly. Or maybe validating first would be sufficient in most cases?

Reply:
Disabling cache invalidation and hooks, waiting for updates to complete, is going to generate a lot of error IMO. The amount of testing to get this into core would far to high and really require a skilled core commiter to comit to this way of doing before we continue with this course of action.